pinot-query-planner/src/main/java/org/apache/pinot/query/planner/stage/CorrelateNode.java

```diff
@@ -0,0 +1,66 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.query.planner.stage;
+
+import java.util.Arrays;
+import java.util.List;
+import org.apache.calcite.rel.core.JoinRelType;
+import org.apache.pinot.common.utils.DataSchema;
+import org.apache.pinot.query.planner.serde.ProtoProperties;
+
+
+public class CorrelateNode extends AbstractStageNode {
```

Comment:
Just curious, if we put decorrelate code the end of toRelation() call, it won't be changed into join node? 

    relRoot = relRoot.withRel(RelDecorrelator.decorrelateQuery(relRoot.rel));

Reply:
this is actually a good idea. we should try and see if this works :-). good call out @61yao

Reply:
Just to complete the thread, decorrelate won't change the Node to Join, but just adding a new ProjectNode before the final reduce.

Here is an example: 
`EXPLAIN PLAN FOR select col1, col3 from a where a.col3 > (select 0.5 * sum(col3) from b where a.col2 = b.col2)`
The plan is 
```
Execution Plan
LogicalProject(col1=[$0], col3=[$2])
  LogicalFilter(condition=[>($2, $7)])
    LogicalCorrelate(correlation=[$cor0], joinType=[left], requiredColumns=[{1}])
      LogicalTableScan(table=[[a]])
      LogicalProject(EXPR$0=[*(0.5:DECIMAL(2, 1), CASE(=($1, 0), null:INTEGER, $0))])
        LogicalAggregate(group=[{}], agg#0=[$SUM0($0)], agg#1=[$SUM0($1)])
          LogicalExchange(distribution=[hash])
            LogicalAggregate(group=[{}], agg#0=[$SUM0($2)], agg#1=[COUNT()])
              LogicalFilter(condition=[=($cor0.col2, $1)])
                LogicalTableScan(table=[[b]])
```
After decorrelate, the plan is:
```
Execution Plan
LogicalProject(col1=[$0], col3=[$2])
  LogicalProject(col1=[$0], col2=[$1], col3=[$2], col4=[$3], col5=[$4], col6=[$5], ts=[$6], EXPR$0=[*(0.5:DECIMAL(2, 1), CASE(=($8, 0), null:INTEGER, $7))])
    LogicalFilter(condition=[>($2, *(0.5:DECIMAL(2, 1), CASE(=($8, 0), null:INTEGER, $7)))])
      LogicalCorrelate(correlation=[$cor0], joinType=[left], requiredColumns=[{1}])
        LogicalTableScan(table=[[a]])
        LogicalAggregate(group=[{}], agg#0=[$SUM0($0)], agg#1=[$SUM0($1)])
          LogicalExchange(distribution=[hash])
            LogicalAggregate(group=[{}], agg#0=[$SUM0($2)], agg#1=[COUNT()])
              LogicalFilter(condition=[=($cor0.col2, $1)])
                LogicalTableScan(table=[[b]])
```